index.js

```diff
@@ -1687,7 +1742,62 @@
   //. //
   //. //   The value at position 1 is not a member of ‘{ x :: FiniteNumber, y :: FiniteNumber }’.
   //. ```
-  function RecordType(fields) {
+  //.
+  //. ##### Extending record types
+  //.
+  //. By using the `parent` argument of the `RecordType` constructor, one can
+  //. define record types which must contain all the same properties of
+  //. another. For example, one could define `Point3D` as a refinement of
+  //. `Point` from the previous example:
+  //.
+  //. ```javascript
+  //. //    Point3D :: Type
+  //. const Point3D = $.RecordType(Point, {z: $.FiniteNumber});
+  //.
+  //. //    p1 :: Point
+  //. const p1 = {x: 0, y: 0};
+  //.
+  //. //    p2 :: Point, Point3D
+  //. const p2 = {x: 3, y: 4, z: 5};
+  //.
+  //. Point.validate(p1);    // => Right({x: 0, y: 0})
+  //. Point3D.validate(p1);  // => Left({propPath: ['z'], value: undefined})
+  //.
+  //. Point.validate(p2);    // => Right({x: 3, y: 4, z: 5})
+  //. Point3D.validate(p2);  // => Right({x: 3, y: 4, z: 5})
+  //.
+  //. dist(p1, p2);
+  //. // => 5
+  //. ```
+  //.
+  //. The `p2` value is a member of `Point3D` as well as `Point`, making it a
+  //. valid argument to `dist`. By default record types allow additional
+  //. properties (which is what allowed `p2` to be a member of `Point`). If one
+  //. wishes for a record type to allow for no additional properties, it can be
+  //. refined as such:
+  //.
+  //. ```javascript
+  //. //    strict :: Type -> Type
+  //. const strict = t => $.NullaryType(
+  //.   'Strict' + t.name,
+  //.   '',
+  //.   t,
+  //.   x => Object.keys(x).length === t.keys.length
+  //. );
```

Comment:
Woah. Using `this` inside a `test` function is new, correct? There are some places where we're passing `t._test` down into a new Type, I'm not sure if the consequences of relying on `this` will include breakage of code relying on the "purity" (*`this`-lessness*?) of `_test`. If it doesn't, then the fact that we can use `this` to refer to the type of a value being `test`ed is actually quite neat.

I do still doubt whether using `UnaryType` for this purpose is useful though. When implementing this as a type refinement, we get the elegance of `strict(User)` being a refinement of `User`. When implementing this as a `UnaryType`, I'm not sure if we gain anything; I believe that adding `Strict(Unknown)` to the environment will not cause `{name: "Bob"} :: User, Strict User`, because `Unknown.type !== 'RECORD'`. This differs from `NonEmpty`, as *its* check does not rely on the type of `types.$1`.